### ISSUE TITLE
BS integration

### ISSUE THREAD

**Issue by ndeutschmann** [madnklo/madnklo#10]
Integrating the soft recoiled the IS is still not working for pp>z+j

Here we document our progress

**Comment by ndeutschmann** [comment]
Generation: we want simpler processes and we can avoid IS gluons both in MNK and in MG5.

madnklo generation
```
set subtraction_currents_scheme colorful_pp
set subtraction_mappings_scheme SoftBeamsRecoilNLO
import model loop_sm
define p = d d~ g
define j = p
define otherflavors= u u~ c c~ s s~ b b~ t t~
generate p p > z j / otherflavors QED=1 --NLO --beam_types=proton@(1,-1,21)
!rm -rf pp_zj_madnklo 
output pp_zj_madnklo
launch pp_zj_madnklo -i
# Then set the flavour cuts in the run card to:
#
##  '(1,-1) (1,-1) > *'
#
```

**Comment by ndeutschmann** [comment]
Reported z+j NLO (*not* LO+NLO) from MG5 by Valentin
`309.8578 +/- 1.97`

**Comment by ndeutschmann** [comment]
Placeholder for re-validation of what had been tested before: Drell-Yan

**Comment by ndeutschmann** [comment]
Moving toward a full test: we integrate ONLY the soft + CS counterterms (no real matrix element) using the reals and the BS configurations. We edit the currents (local and BS) to cut the soft divergence and furthemore edit the ME7_integrands to select only the relevant local CTs

**Comment by ndeutschmann** [comment]
[beam_factorization.py](https://github.com/madnklo/madnklo/files/2658342/beam_factorization.txt)
[local_currents.py](https://github.com/madnklo/madnklo/files/2658345/local_currents.txt)
[ME7_integrands.py](https://github.com/madnklo/madnklo/files/2658347/ME7_integrands.txt)

First tests with low stats are encouraging (this worked at permille level before pull+clean to 608). More stats being generated

**Comment by ValentinHirschi** [comment]
One can first compare ```MadNkLO``` against ```MG5aMC``` when including only collinear configurations by forcing to include only the gluon quark contribution in the initial states.
This can be done in ```MG5aMC``` with:
```
# c     This will be called for any PDG code, but we only support up to 7
# add the following lines:
#
##  if(iabs(ipart).ne.0) then
##    pdg2pdf=1d-50
##    return
##  endif
```
and in ```MadNkLO``` with:
```
# Then set the flavour cuts in the run card to:
#
##  '21 21 > *'
#
```
The LO result is of course zero in this case, and we then only have the NLO contribution (which can therefore be interpreted as pure NLO), which amounts to the following in ```MG5aMC```:
```
MG5_aMC NLO result for the g g > * contribution:
<TBD>
```
While in ```MadNkLO``` we find:
```
MadNkLO NLO result for the g g > * contribution:
R   = <TBD>
BF1 = <TBD>
BF2 = <TBD>
Sum = <TBD>
```

**Comment by ndeutschmann** [comment]
# Results without flavor cuts for `pp_zj_madnklo_all_flavours` (MadNkLO)

R = -3942 +/- 95
BS = 

details:
revno 610
modified beam_subtraction current: diff below
```
501c501
<                         kernel = EpsilonExpansion({0:HE.integrated_bs_counterterm_finite(dipole_invariant,xi)})
---
>                         kernel = EpsilonExpansion({0:HE.integrated_bs_counterterm_finite(dipole_invariant,xi)})*xi**2
```
Generation is 
```
set subtraction_currents_scheme colorful_pp
set subtraction_mappings_scheme SoftBeamsRecoilNLO
import model loop_sm
define p = d d~ u u~ c c~ s s~ g
define j = p
generate p p > z j QED=1 --NLO
!rm -rf pp_zj_madnklo_all_flavours 
output pp_zj_madnklo_all_flavours
launch pp_zj_madnklo_all_flavours -i
```

**Comment by ValentinHirschi** [comment]
Results obtained for the flavour configuration `(1,-1) (1,-1) > *`:
```
MG5aMC:
LO:
1.306e+03 +- 8.2e+00 pb
LO+NLO:
```
vs MadNkLO_ISR, revno 610, with modified beam_subtraction current: diff below
```
501c501
<                         kernel = EpsilonExpansion({0:HE.integrated_bs_counterterm_finite(dipole_invariant,xi)})
---
>                         kernel = EpsilonExpansion({0:HE.integrated_bs_counterterm_finite(dipole_invariant,xi)})*xi**2
```
giving
```
LO:
1.31662e+03 +/- 1.08e+00 [pb]
pure NLO:
R   = <TBD>
V   = <TBD>
BS  = <TBD> 
BF1 = <TBD>
BF2 = <TBD>
```

**Comment by ndeutschmann** [comment]
Latest results for pp_zj no grouping 

[cross_sections.txt](https://github.com/madnklo/madnklo/files/2754899/cross_sections.txt)

[run_card.txt](https://github.com/madnklo/madnklo/files/2754900/run_card.txt)

**Comment by ValentinHirschi** [comment]
So in conclusion we have `MadFKS` result being:
`NLO-LO = (1.495e+04 +- 1.1e+02) - (1.130e+04 +- 3.2e+01) = 3650 +- 130`

Vs `MadNkLO` result being:
`BF1+BF2+BS+R+V=1891+1877+4388+-4219+1304 = 5241 +- 200`

**Comment by ValentinHirschi** [comment]
One more check one can do when doing the bulk test for the soft subtraction using the higher PS point. We have the *analytic* expression for the mapped momenta as a function of the unmapped ones with Vittorio's mapping. Therefore we could then *numerically* check the value of these mapped quantities obtained directly from the lower_PS_point computed by the mapping performed in `mappings.py` against the one obtained from the analytic expression of these mapped quantities with the "higher_PS_point" in input.

**Comment by ndeutschmann** [comment]
This test highlighted a possible issue with the mapping!
 
Here's a report of the results for `p p > z j` in `S((4, 21, f)) @ 0 loops`

```
Higher PS point:
#     E                         p_x                       p_y                       p_z                       M
--------------------------------------------------------------------------------------------------------------------------------------
1     6.2732045896392665e+03    0.0000000000000000e+00    0.0000000000000000e+00    6.2732045896392665e+03    0.0000000000000000e+00
2     6.2732045896392665e+03    0.0000000000000000e+00    0.0000000000000000e+00   -6.2732045896392665e+03    0.0000000000000000e+00
3     3.5974567949353082e+03    2.8154873129977312e+02    2.4540896920929340e+03    2.6137242079356643e+03    9.1188000000010149e+01
4     3.0552978846124502e+03   -2.9782914521191287e+02   -6.0055026216020008e+02    2.9808526207444356e+03    0.0000000000000000e+00
5     5.8936544997307738e+03    1.6280413912139750e+01   -1.8535394299327340e+03   -5.5945768286800994e+03    6.1035156250000000e-05
--------------------------------------------------------------------------------------------------------------------------------------
Sum   1.8189894035458565e-12    0.0000000000000000e+00    0.0000000000000000e+00    0.0000000000000000e+00
```

```
Lower PS point
#     E                         p_x                       p_y                       p_z                       M
--------------------------------------------------------------------------------------------------------------------------------------
1     4.4929485987495109e+03    0.0000000000000000e+00    0.0000000000000000e+00    4.4929485987495109e+03    0.0000000000000000e+00
2     4.4929485987495109e+03    0.0000000000000000e+00    0.0000000000000000e+00   -4.4929485987495109e+03    0.0000000000000000e+00
3     4.4934112821722774e+03    1.5113279711599540e+02    2.1911156859451894e+03    3.9190050558592006e+03    9.1187999999862058e+01
5     4.4924859153267371e+03   -1.5113279711599537e+02   -2.1911156859451894e+03   -3.9190050558592002e+03    1.4314012938914275e-04
--------------------------------------------------------------------------------------------------------------------------------------
Sum   7.2759576141834259e-12   -2.8421709430404007e-14    0.0000000000000000e+00   -4.5474735088646412e-13
```

Let us look at the initial-final dipole (1,5)
```
pa = higher_PS_point[1]
pb = higher_PS_point[5]
pa_tilde = lower_PS_point[1]
pb_tilde = lower_PS_point[5]
pS = higher_PS_point[4]
```

We also have `xi = 0.716212668429`. (`xi = math.sqrt(1 - 2. * pS.dot(Q) / Q.square())`)

The mapping provides a relation between the invariant s = 2 pa.pb and the invariants written in terms of momenta from the lower-multiplicity phase space point. This relation can be found in Vittorio's handwritten notes about his mapping or using the observation that both initial- and final-state mapped momenta are related to those of the Colorful NLO soft mapping by a rescaling of ξ.

All in all we can use the following relation:
![screen shot 2019-01-29 at 12 57 18](https://user-images.githubusercontent.com/12328649/51906919-a2e5a800-23c5-11e9-83d1-59765fbde3bb.png)
Vittorio has a slightly different formula which he proved (and we cross-checked) is equivalent to that found in colorful.

We define the following results, which should all be equal:
```
# Straight from the higher_PS_point
sai = 2.*pa.dot(pb)
# From Vittorio's formula
saitgtV = 2*pa.dot(pb_tilde) - 1/(1+xi) *2.*pa.dot(pS)*2.*pb_tilde.
dot(Q)/Q.square()-1/xi/(1+xi)**2*2.*pa.dot(pS)*2.*pb_tilde.dot(pS)/Q.squa
re()+1/xi/(1+xi)* 2.* pb_tilde.dot(pS)
# From the colorful formula
sait = 2.*pa_tilde.dot(pb_tilde) #s_{\tilde{a}\tilde{i}}
satQ = 2.*pa_tilde.dot(Q) #s_{\tilde{a}Q}
sitQ=2.*pb_tilde.dot(Q) #s_{\tilde{i}Q}
satr = 2.*pa_tilde.dot(pS) #s_{\tilde{a}r}
sitr = 2.*pb_tilde.dot(pS) #s_{\tilde{i}r}
saitgt = sait/xi + 1/xi**2/(1+xi)*satQ*sitr/Q.square()-1/xi/(1+xi)*satr*sitQ/Q.square()-1/xi**2/(1+xi)**2*satr*sitr/Q.square()
```
We find the following result:
```
sai = 157412383
saitgt = 144136050
saitgtV = 144136050
```
(This should give us confidence that we did not make a typo when implementing either formula since they match, and that there is an issue with the mapping)

**Comment by ndeutschmann** [comment]
Latest results for pp_zj (Vittorio_softs 36aa1d29488a867f49222ecd1005f8eca5367f27)
# MadNkLO Result
[cross_sections.txt](https://github.com/madnklo/madnklo/files/2821815/cross_sections.txt)

**Comment by ValentinHirschi** [comment]
Latest results from git rev `b08d0989f2a3b614274a9c513e679b1d4e323b75` of the `master` branch.

# Run setup

Common or similar across all processes.

[generate_pp_epemj_no_a_colorful_pp_ungrouped.txt](https://github.com/madnklo/madnklo/files/3047702/generate_pp_epemj_no_a_colorful_pp_ungrouped.txt)
[generate_pp_epemj_no_a_colorful_ungrouped.txt](https://github.com/madnklo/madnklo/files/3047704/generate_pp_epemj_no_a_colorful_ungrouped.txt)
[run_card.txt](https://github.com/madnklo/madnklo/files/3047714/run_card.txt)
[param_card.txt](https://github.com/madnklo/madnklo/files/3047715/param_card.txt)

# d d~ > z | d~ d > z

```
MadFKS LO       : 1.750e+04 +- 6.4e+00 pb
MadNkLO         : 1.74764e+04 +/- 1.57e+00 [pb]
MadFKS pure NLO: 5.084e+03 +- 3.5e+00 pb


pure-NLO:
5.084e+03 +- 3.5e+00 pb

Super-Basic:
set_integrator_options VEGAS3 --n_iterations_survey=10 --n_points_survey=1000 --n_iterations_refine=5 --n_points_refine=5000 --nb_core=48

Basic R:
set_integrator_options VEGAS3 --n_iterations_survey=10 --n_points_survey=10000 --n_iterations_refine=5 --n_points_refine=50000 --nb_core=48

Advanced R:
set_integrator_options VEGAS3 --n_iterations_survey=10 --n_points_survey=50000 --n_iterations_refine=5 --n_points_refine=500000 --nb_core=48

**Comment by ValentinHirschi** [comment]
Latest results from git rev `8d7d9a81329bd7fb3869975eaacec913011bf564` of the `counterterm_refactoring` branch.

# Run setup

Common or similar across all processes and same as previous post of April 5th.
The generation script only involves specifying p p > d d~.
Note: the initial state read `d d~` but it also includes `d~ d`.

# d d~ > z

```
MadFKS LO        : 1.750e+04 +- 6.4e+00 pb
MadNkLO          :  1.74780e+04 +/- 4.21e-02 [pb]
MadFKS pure NLO  : 5.084e+03 +- 3.5e+00 [pb]
MadNkLO pure NLO : 5.081.8  +/- 0.5 [pb]

MadNkLO B   :  1.74780e+04 +/- 4.21e-02 [pb]
MadNkLO R   : -1.36754e+02 +/- 2.34e-01 [pb]
MadNkLO BF1 :  2.70982e+03 +/- 3.12e-01 [pb]
MadNkLO BF2 :  2.70976e+03 +/- 3.33e-01 [pb]
MadNkLO BS  :  0.0 [pb]
MadNkLO V   : -2.01042e+02 +/- 4.86e-04 [pb]
```

# d d~ > z j, grouped, ptj>20 GeV:

```
MadFKS LO        : 1.313e+03 +- 9.6e-01 [pb]
MadNkLO          : 1.31334e+03 +/- 3.80e+00 [pb]
MadFKS pure NLO  : 310.0 +- 2.0 [pb]
MadNkLO pure NLO : 322.6 +- 6  [pb] (previous wrong impl. yielded : 776.1 +- 18)

MadNkLO B   :  1.31334e+03 +/- 3.80e+00 [pb]
MadNkLO R   : -7.11466e+02 +/- 2.60e+01 [pb] (--n_iterations_survey=10 --n_points_survey=100000 --n_iterations_refine=3 --n_points_refine=1000000)
MadNkLO R   : -7.18608e+02 +/- 6.09e+00 [pb] (--n_iterations_survey=10 --n_points_survey=100000 --n_iterations_refine=3 --n_points_refine=1000000)
MadNkLO BF1 :  2.17827e+02 +/- 8.61e-01 [pb]
MadNkLO BF2 :  1.73178e+02 +/- 6.00e-01 [pb]
MadNkLO BS  :  6.22238e+02 +/- 6.73e-01 [pb]
MadNkLO V   : 2.79330e+01 +/- 4.79e-02 [pb]
```

# d d~ > z j, ungrouped, ptj>20 GeV:

```
MadFKS LO        : 1.313e+03 +- 9.6e-01 [pb]
MadNkLO          : 1.33264e+03 +/- 1.78e+01 [pb]
MadFKS pure NLO  : 310.0 +- 2.0 [pb]
MadNkLO pure NLO : 322.6 +- 6  [pb] (previous wrong impl. yielded : 776.1 +- 18)

MadNkLO B   :  1.31334e+03 +/- 3.80e+00 [pb]
MadNkLO R   : -7.37607e+02 +/- 1.83e+01 [pb] (--n_points_survey=100000 --n_iterations_refine=3 --n_points_refine=1000000)
itn   integral        average         chi2/dof        Q
-------------------------------------------------------
  1   -719.6(8.1)     -719.6(8.1)         0.00     1.00
  2   -701(24)        -711(13)            0.51     0.47
  3   -730.3(9.4)     -717.1(9.1)         0.87     0.42
MadNkLO BF1 :  1.96006e+02 +/- 7.18e-01 [pb]
MadNkLO BF2 :  1.95999e+02 +/- 7.09e-01 [pb]
MadNkLO BS  :  6.23195e+02 +/- 2.46e+00 [pb]
MadNkLO V   : 2.80711e+01 +/- 1.86e-01 [pb]
```

# d d~ > z j, ungrouped, ptj>100 GeV:

```
MadFKS LO        : 3.589e+01 +- 7.3e-02 pb [pb]
MadNkLO          : 3.56855e+01 +/- 3.43e-01 [pb]
MadFKS pure NLO  : 7.94 +- 0.76 [pb]
MadNkLO pure NLO : 7.712 +- 0.16 [pb] (old MadNkLO result: 9.36 +- 0.6)

MadNkLO B   :  3.56855e+01 +/- 3.43e-01 [pb]
MadNkLO R   : -1.33503e+01 +/- 6.40e-01 [pb]
MadNkLO BF1 :  5.84964e+00 +/- 1.47e-01 [pb]
MadNkLO BF2 :  5.71277e+00 +/- 7.56e-02 [pb]
MadNkLO BS  :  1.52097e+01 +/- 3.73e-01 [pb]
MadNkLO V   : -5.47278e+00 +/- 4.68e-02 [pb]
```

**Comment by ValentinHirschi** [comment]
Ok, the higher stat results came in (3.5 days ... ) and are spot on (`310.0 +- 2.0` vs `308.2 +- 4.6`) so we can finally close this issue, merge counterterm refactoring to master and move on with our lives.

**Comment by ndeutschmann** [comment]
Updating the status in the `full_nnlo` branch where we now have convolution support for electron beams. We re-validated `e+ e- >  d d~`  at NLO

## Setup
- commit `4d237cd63249ecc923d6d28ebb1e1ef91033cb03` 
- integrator is `VEGAS` with settings 
``` 
'VEGAS' : (pyCubaIntegrator.pyCubaIntegrator,
                  { 'algorithm' : 'Vegas',
                    'verbosity' : integrator_verbosity,
                    'seed'      : 42,
                    'target_accuracy' : 1.0e-2,
                    'target_accuracy_survey' : 5.0e-3,
                    'n_start'           : 30000,
                    'n_start_survey'    : 30000,
                    'n_increase'        : 5000,
                    'n_increase_survey' : 5000,
                    'n_batch'           : 1000,
                    'max_eval'          : int(1e7),
                    'max_eval_survey'   : int(1e7),
                    'min_eval'          : 0}),
```

- process generation:

```
import model loop_sm
set subtraction_scheme colorful_pp
set group_subprocesses True
define p = d d~ g
define j = p
define otherflavors= u u~ c c~ s s~ b b~ t t~
generate e+ e- > j j / otherflavors z QED=2 --NLO --beam_one_type=lepton@(-11,) --beam_two_type=lepton@(11,)
!rm -rf epem_jj_colorful
output epem_jj_colorful
launch epem_jj_colorful
```

- relevant run card settings: use `lpp=1` to allow beam recoil and remove jet cuts

```
#*********************************************************************
# Collider type and energy                                           *
# lpp: 0=No PDF, 1=proton, -1=antiproton, 2=photon from proton,      *
#                                         3=photon from electron     *
#*********************************************************************
     1        = lpp1    ! beam 1 type
     1        = lpp2    ! beam 2 type
     500.0     = ebeam1  ! beam 1 total energy in GeV
     500.0     = ebeam2  ! beam 2 total energy in GeV
#*********************************************************************
[...]
#*********************************************************************
# Minimum and maximum pt's (for max, -1 means no cut)                *
#*********************************************************************
 -1  = ptj       ! minimum pt for the jets
[...]
#*********************************************************************
# Maximum and minimum absolute rapidity (for max, -1 means no cut)   *
#*********************************************************************
 -1 = etaj    ! max rap for the jets
[...]
#*********************************************************************
# Minimum and maximum DeltaR distance                                *
#*********************************************************************
 -1 = drjj    ! min distance between jets
```

- PDF is set by overriding `madgraph.integrator.ME7_integrands.ME7Integrand.get_pdfQ2`:
```
    def get_pdfQ2(self, pdf, pdg, x, scale2):
        """ Call the PDF and return the corresponding density."""

        return 1. if x > 0.5 else 0.
```
(note that otherwise, without masses or cuts we can go all the way to zero energies, which leads to issues)


For a general process we would need to validate with MG5 but for `e+ e- > hadrons` we know that the NLO correction is a_s/pi locally in Bjorken x's so as long as we integrate the NLO and the LO with the same PDFs we expect the ratio to be a_s/pi. 

## Results

```
{
    "B": {
        "Cross section": 0.014901073106384494,
        "MC uncertainty": 5.8745515280783016e-05,
        "name": "B",
        "order": "LO",
        "timestamp": "2020:02:10:16:16:27",
        "unit": "pb"
    },
    "BS": {
        "Cross section": -0.000746968287563887,
        "MC uncertainty": 7.456723298631915e-06,
        "name": "BS",
        "order": "NLO",
        "timestamp": "2020:02:11:12:40:41",
        "unit": "pb"
    },
    "R": {
        "Cross section": -0.0016662175139998063,
        "MC uncertainty": 6.213544576941782e-06,
        "name": "R",
        "order": "NLO",
        "timestamp": "2020:02:10:16:19:56",
        "unit": "pb"
    },
    "V": {
        "Cross section": 0.002985044155105842,
        "MC uncertainty": 1.1768142857536976e-05,
        "name": "V",
        "order": "NLO",
        "timestamp": "2020:02:11:11:52:24",
        "unit": "pb"
    }
}
```

Which yields

```
LO = 1.4901e-02 +/- 5.8746e-05
a_s/pi * LO = 5.5971e-04 +/- 2.2066e-06 (a_s = 0.118)
NLO = 5.7186e-04 +/- 1.5255e-05
```

**Comment by ndeutschmann** [comment]
We will keep this issue open until we fully validate `colorful_pp` at NLO in the `full_nnlo` branch

**Comment by ndeutschmann** [comment]
# Validation of Drell Yan

**TL;DR**
As we should, we get contribution-by-contribution agreement.

**Reminder**
We compare with values of https://github.com/madnklo/madnklo/issues/10#issuecomment-491855817

## Code version
`49567c0a51a4d8f45150a48565bb42e90540b6a7`
with the following modification:

```
diff --git a/madgraph/integrator/ME7_integrands.py b/madgraph/integrator/ME7_integrands.py
index 42f5617bd..6be6460a0 100644
--- a/madgraph/integrator/ME7_integrands.py
+++ b/madgraph/integrator/ME7_integrands.py
@@ -1722,18 +1722,22 @@ class ME7Integrand(integrands.VirtualIntegrand):

         return self.run_card['scale'], self.run_card['dsqrt_q2fact1'], self.run_card['dsqrt_q2fact2']

-    def get_pdfQ2(self, pdf, pdg, x, scale2):
+    def get_pdfQ2(self, pdf, pdg, x, scale2,nonQCD_cutoff=0.1):
         """ Call the PDF and return the corresponding density."""

         # For testing, one can easily substitute PDFs with a 1./x distribution by uncommenting the
         # line below
-        #return 1./x
+        #return 1/x

         if pdf is None:
             return 1.

         if pdg not in [21,22] and abs(pdg) not in range(1,7):
-            return 1.
+            # This should be called only when debugging as long as we have no QED corrections:
+            # non-partonic beam that has a "beam type" set to 1 in the run_card
+            # typical use case: colorful_pp scheme tests with e+ e- initial states
+            return 1. if x > nonQCD_cutoff else 0.
+

         if (pdf, pdg, x, scale2) in self.PDF_cache:
             return self.PDF_cache[(pdf, pdg, x, scale2)]
@@ -2047,7 +2051,8 @@ class ME7Integrand(integrands.VirtualIntegrand):
             process['beam_factorization'], PS_point, process, mu_r, mu_f1, mu_f2, xb_1, xb_2, xi1, xi2 )

         # Log the abort status in the event
-        convolved_event.abort = abort_status
+        if convolved_event is not None:
+            convolved_event.abort = abort_status

         return convolved_event
```
* First part should be irrelevant
* Second part is committed to `c6054f0e3a7cf4a979dee77a4712a78436da16a3`

## Generation

```
import model loop_sm
set subtraction_scheme colorful_pp
set group_subprocesses False
define p = d d~ g
define j = p
define otherflavors= u u~ c c~ s s~ b b~ t t~
generate p p > z / otherflavors QED=1 --NLO --beam_types=proton@(1,-1,21)
!rm -r CHECK_49567c0a5_pp_z_1f_ungrouped
output CHECK_49567c0a5_pp_z_1f_ungrouped
```

## Run card

```
# Collider type and energy                                           *
# lpp: 0=No PDF, 1=proton, -1=antiproton, 2=photon from proton,      *
#                                         3=photon from electron     *
#*********************************************************************
     1        = lpp1    ! beam 1 type
     1        = lpp2    ! beam 2 type
     6500.0     = ebeam1  ! beam 1 total energy in GeV
     6500.0     = ebeam2  ! beam 2 total energy in GeV
#*********************************************************************
# PDF CHOICE: this automatically fixes also alpha_s and its evol.    *
#*********************************************************************
     lhapdf    = pdlabel     ! PDF set
     90900    = lhaid     ! if pdlabel=lhapdf, this is the lhapdf number
#*********************************************************************
# Renormalization and factorization scales                           *
#*********************************************************************
 True = fixed_ren_scale  ! if .true. use fixed ren scale
 True        = fixed_fac_scale  ! if .true. use fixed fac scale
 91.188  = scale            ! fixed ren scale
 91.188  = dsqrt_q2fact1    ! fixed fact scale for pdf1
 91.188  = dsqrt_q2fact2    ! fixed fact scale for pdf2
 -1 = dynamical_scale_choice ! Choose one of the preselected dynamical choices

[...]

#*********************************************************************
# Standard Cuts                                                      *
#*********************************************************************
 '(1,-1) (1,-1) > *' = flavor_cuts ! function string, or keyword: 'hardcoded'
#*********************************************************************
# Minimum and maximum pt's (for max, -1 means no cut)                *
#*********************************************************************
 20.0  = ptj       ! minimum pt for the jets
 10.0  = pta       ! minimum pt for the photons
 10.0  = ptl       ! minimum pt for the charged leptons
#*********************************************************************
# Maximum and minimum absolute rapidity (for max, -1 means no cut)   *
#*********************************************************************
 -1 = etaj    ! max rap for the jets
 2.5 = etaa    ! max rap for the photons
 2.5 = etal    ! max rap for the charged leptons
#*********************************************************************
# Minimum and maximum DeltaR distance                                *
#*********************************************************************
 0.4 = drjj    ! min distance between jets
 0.4 = drll    ! min distance between leptons
 0.4 = draa    ! min distance between gammas
 0.4 = draj    ! min distance between gamma and jet
 0.0 = drjl    ! min distance between jet and lepton
 0.4 = dral    ! min distance between gamma and lepton
```

## Results

```
{
    "B": {
        "Cross section": 17468.524160492478,
        "MC uncertainty": 3.1426462275018054,
        "name": "B",
        "order": "LO",
        "timestamp": "2020:05:11:17:03:53",
        "unit": "pb"
    },
    "BF1": {
        "Cross section": 2708.5587199850484,
        "MC uncertainty": 3.1416918429681933,
        "name": "BF1",
        "order": "NLO",
        "timestamp": "2020:05:11:16:58:03",
        "unit": "pb"
    },
    "BF2": {
        "Cross section": 2707.625366256493,
        "MC uncertainty": 3.3595726756919992,
        "name": "BF2",
        "order": "NLO",
        "timestamp": "2020:05:11:16:58:25",
        "unit": "pb"
    },
    "BS": {
        "Cross section": 0.0,
        "MC uncertainty": 1.4142135623730952e-129,
        "name": "BS",
        "order": "NLO",
        "timestamp": "2020:05:11:17:03:36",
        "unit": "pb"
    },
    "R": {
        "Cross section": -135.09904517442988,
        "MC uncertainty": 1.1500383398838885,
        "name": "R",
        "order": "NLO",
        "timestamp": "2020:05:11:16:59:25",
        "unit": "pb"
    },
    "V": {
        "Cross section": -201.04600379393986,
        "MC uncertainty": 0.04846840925836935,
        "name": "V",
        "order": "NLO",
        "timestamp": "2020:05:11:17:03:19",
        "unit": "pb"
    }
}
```
